File: src/main/java/cn/labzen/network/onvif/Onvif.java
package cn.labzen.network.onvif;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.net.InetAddress;
import java.net.InterfaceAddress;
import java.net.NetworkInterface;
import java.net.SocketException;
import java.util.*;
import java.util.function.Function;
import java.util.stream.Stream;

public final class Onvif {

  private static final Logger LOGGER = LoggerFactory.getLogger(Onvif.class);

  private Onvif() {
  }

  public static Discoverer create(int timeout) {
    return new Discoverer(timeout);
  }

  /**
   * 获取当前设备上的接口地址列表
   */
  public static List<InetAddress> interfaceAddresses() {
    try {
      return addresses(InterfaceAddress::getAddress);
    } catch (SocketException e) {
      LOGGER.warn("on Onvif.interfaceAddresses()", e);
      return null;
    }
  }

  /**
   * 获取当前设备上所有广播地址的列表
   */
  public static List<InetAddress> broadcastAddresses() {
    try {
      return addresses(InterfaceAddress::getBroadcast);
    } catch (SocketException e) {
      LOGGER.warn("on Onvif.broadcastAddresses()", e);
      return null;
    }
  }

  private static List<InetAddress> addresses(Function<InterfaceAddress, InetAddress> function) throws SocketException {
    Enumeration<NetworkInterface> interfaceEnumeration = NetworkInterface.getNetworkInterfaces();
    ArrayList<NetworkInterface> interfaces = Collections.list(interfaceEnumeration);
    return interfaces.stream().flatMap(ni -> {
      try {
        if (ni.isLoopback() || !ni.isUp()) {
          return Stream.empty();
        }
        return ni.getInterfaceAddresses().stream().map(function);
      } catch (SocketException e) {
        return Stream.empty();
      }
    }).filter(Objects::nonNull).toList();
  }
}

File: src/main/kotlin/cn/labzen/network/onvif/Onvif.kt
package cn.labzen.network.onvif

import org.slf4j.LoggerFactory
import java.net.Inet4Address
import java.net.NetworkInterface
import java.net.SocketException

@Suppress("unused")
object Onvif {

  private val logger = LoggerFactory.getLogger(Onvif::class.java)

  @JvmStatic
  @JvmOverloads
  fun create(timeout: Int = 5000) =
    Discoverer(timeout)

  /**
   * 获取此设备上的接口地址列表
   */
  fun interfaceAddresses() =
    try {
      NetworkInterface.getNetworkInterfaces().toList().flatMap { ni ->
        if (ni.isLoopback || !ni.isUp) {
          emptyList()
        } else {
          ni.interfaceAddresses.map { a -> a.address }.toList()
        }
      }.filterNotNull()
    } catch (e: SocketException) {
      logger.warn("on Onvif.interfaceAddresses()", e)
      emptyList()
    }

  /**
   * 获取此设备上所有广播地址的列表
   */
  fun broadcastAddresses() =
    try {
      NetworkInterface.getNetworkInterfaces().toList().flatMap { ni ->
        if (ni.isLoopback || !ni.isUp) {
          emptyList()
        } else {
          ni.interfaceAddresses.map { a -> a.broadcast }.toList()
        }
      }.filterNotNull()
    } catch (e: SocketException) {
      logger.warn("on Onvif.broadcastAddresses()", e)
      emptyList()
    }

  /**
   * 获取此设备本地IP
   */
  fun localIp() =
    try {
      NetworkInterface.getNetworkInterfaces().toList().firstNotNullOfOrNull { ni ->
        ni.inetAddresses.toList().findLast { inet -> inet.isLoopbackAddress && inet is Inet4Address }?.hostAddress
      }
    } catch (e: SocketException) {
      logger.warn("on Onvif.broadcastAddresses()", e)
      null
    }
}
